src/vfs2.c

```diff
@@ -928,23 +928,45 @@ static int compare_wal_headers(struct wal_hdr a,
 	return SQLITE_OK;
 }
 
-static int read_wal_hdr(sqlite3_file *wal,
-			sqlite3_int64 *size,
-			struct wal_hdr *hdr)
+static bool wal_hdr_is_valid(const struct wal_hdr *hdr)
+{
+	/* TODO(cole) Add other validity constraints. */
+	struct cksums sums_found = {};
+	const uint8_t *p = (const uint8_t *)hdr;
+	size_t len = offsetof(struct wal_hdr, cksum1);
+	update_cksums(p, len, &sums_found);
+	struct cksums sums_expected = { ByteGetBe32(hdr->cksum1),
+					ByteGetBe32(hdr->cksum2) };
+	return cksums_equal(sums_expected, sums_found);
+}
+
+/**
+ * Read the header of the given WAL file and detect corruption.
+ *
+ * If the file contains a valid header, returns this header in `hdr` and the
+ * size of the file in `size`. If the file is too short to contain a header,
+ * returns the size only. If the file can't be read, or it contains an invalid
+ * header, returns an error.
+ */
+static int try_read_wal_hdr(sqlite3_file *wal,
+			    sqlite3_int64 *size,
+			    struct wal_hdr *hdr)
 {
 	int rv;
 
 	rv = wal->pMethods->xFileSize(wal, size);
 	if (rv != SQLITE_OK) {
 		return rv;
 	}
-	if (*size >= (sqlite3_int64)sizeof(struct wal_hdr)) {
-		rv = wal->pMethods->xRead(wal, hdr, sizeof(*hdr), 0);
-		if (rv != SQLITE_OK) {
-			return rv;
-		}
-	} else {
-		*hdr = (struct wal_hdr){};
+	if (*size < (sqlite3_int64)sizeof(*hdr)) {
```

Comment:
Yes how can we get to that state, is that a valid state at all?

Reply:
I wrote this having in mind that a WAL header write could result in only some of the bytes making it to disk---that was the motivation for treating a WAL with an incomplete header as though it were empty.

However, this is likely to be quite rare, and probably it would be better to just treat such a WAL as corrupt (since a short file could result from other things going wrong as well that we wouldn't want to ignore---hardware failure or a dqlite bug or etc).

Reply:
Interestingly, SQLite itself will just ignore a WAL whose header is corrupt. I'm not sure if that policy makes sense in dqlite's case, we probably want to fail early instead because raft will be counting on certain data being present in the WALs. But we can relax this later if we decide that would be a good idea.